xm-commons-lep-groovy/src/main/java/com/icthh/xm/commons/lep/groovy/GroovyLepEngine.java

```diff
@@ -0,0 +1,196 @@
+
+@Slf4j
+public class GroovyLepEngine extends LepEngine {
+
+    public static final String LEP_PREFIX = "lep://";
+    public static final String COMMONS = "commons/lep";
+    public static final String COMMONS_SCRIPT = "/Commons$$";
+    private final String appName;
+    private final String tenant;
+    private final LepStorage leps;
+    private final GroovyScriptEngine gse;
+    private final LoggingWrapper loggingWrapper;
+
+    private final Map<String, GroovyFileParser.GroovyFileMetadata> lepMetadata = new ConcurrentHashMap<>();
+
+    public GroovyLepEngine(String appName,
+                           String tenant,
+                           LepStorage leps,
+                           LoggingWrapper loggingWrapper,
+                           ClassLoader classLoader,
+                           Map<String, GroovyFileParser.GroovyFileMetadata> lepMetadata,
+                           LepResourceConnector lepResourceConnector,
+                           boolean isWarmupEnabled) {
+        this.appName = appName;
+        this.tenant = tenant;
+        this.leps = leps;
+        this.loggingWrapper = loggingWrapper;
+        this.gse = buildGroovyEngine(classLoader, lepResourceConnector);
+        this.lepMetadata.putAll(lepMetadata);
+        if (isWarmupEnabled) {
+            warmupScripts();
+        } else {
+            log.warn("Warmup lep script disabled");
+        }
+    }
+
+    protected GroovyScriptEngine buildGroovyEngine(ClassLoader classLoader, LepResourceConnector lepResourceConnector) {
+        var gse = new GroovyScriptEngine(lepResourceConnector, classLoader);
+        CompilerConfiguration config = gse.getConfig();
+        config.setRecompileGroovySource(true);
+        config.setMinimumRecompilationInterval(50);
+        gse.setConfig(config);
+        gse.getGroovyClassLoader().setShouldRecompile(true);
+        return gse;
+    }
+
+    private void warmupScripts() {
+        StopWatch stopWatch = StopWatch.createStarted();
+        log.info("Start warmup lep scripts");
+        this.leps.forEach(lep -> {
+            try {
+                if (!isCommonsClass(lep.getPath())) {
+                    if (lepMetadata.containsKey(lep.metadataKey()) && lepMetadata.get(lep.metadataKey()).isScript()) {
+                        StopWatch warmUpTime = StopWatch.createStarted();
+                        log.trace("START | Warmup lep {}", lep.getPath());
+                        Class<?> scriptClass = gse.loadScriptByName(LEP_PREFIX + lep.getPath());
+                        InvokerHelper.getMetaClass(scriptClass); // build metaclass
+                        log.trace("STOP | Warmup lep {}, time: {}", lep.getPath(), warmUpTime.getTime(MILLISECONDS));
+                    }
+                }
+            } catch (Throwable e) {
+                log.error("Error create script {}", lep.getPath(), e);
+            }
+        });
+        log.info("End warmup lep scripts | time {}ms", stopWatch.getTime(MILLISECONDS));
```

Comment:
suggest log statement: 
```
        log.info("Stop warm-up LEP scripts, count = {}, time = {} ms, ", leps.size(), stopWatch.getTime(MILLISECONDS));
```

Reply:
size not available, because storage can be file system, etc